File: app/src/test/java/com/originb/inkwisenote/io/sql/NoteTermFrequencyDaoTest.java
package com.originb.inkwisenote.io.sql;

import androidx.room.Room;
import com.originb.inkwisenote.common.NotesDatabase;
import com.originb.inkwisenote.modules.ocr.data.NoteTermFrequencyDao;
import com.originb.inkwisenote.modules.ocr.data.NoteTermFrequency;
import com.originb.inkwisenote.modules.ocr.data.TermOccurrence;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import java.util.*;

import static org.junit.Assert.*;

@RunWith(RobolectricTestRunner.class)
public class NoteTermFrequencyDaoTest {
    private NoteTermFrequencyDao noteTermFrequencyDao;
    private NotesDatabase db;

    @Before
    public void setUp() {
        db = Room.inMemoryDatabaseBuilder(RuntimeEnvironment.getApplication(),
                        NotesDatabase.class)
                .allowMainThreadQueries() // Use cautiously, only for tests
                .build();
        noteTermFrequencyDao = db.noteTermFrequencyDao();
    }

    @After
    public void tearDown() {
        db.close();
    }

    @Test
    public void testInsertAndReadTermFrequencies() {
        // Prepare test data
        Long noteId = 1L;
        List<NoteTermFrequency> noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 2));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "banana", 1));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "orange", 3));

        // Insert data
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        List<NoteTermFrequency> termFrequencies = noteTermFrequencyDao.getAllTermFrequencies();
        assertEquals(noteTermFrequencyList.size(), termFrequencies.size());

        for (NoteTermFrequency expected : noteTermFrequencyList) {
            boolean areTheyEqual = termFrequencies.stream().anyMatch(actual ->
                    actual.getNoteId() == expected.getNoteId()
                            && actual.getTerm().equals(expected.getTerm())
                            && actual.getTermFrequency() == expected.getTermFrequency()
            );
            assertTrue(areTheyEqual);
        }
    }

    @Test
    public void testGetTermOccurrences() {
        // Insert test data for multiple notes
        Long noteId = 1L;
        List<NoteTermFrequency> noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 2));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "banana", 1));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        noteId = 2L;
        noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 1));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "orange", 3));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        // Test getTermOccurrences
        Set<String> termsToQuery = new HashSet<>();
        termsToQuery.add("apple");
        termsToQuery.add("banana");
        termsToQuery.add("orange");

        List<TermOccurrence> occurrences = noteTermFrequencyDao.getTermOccurrences(termsToQuery);

        assertEquals(occurrences.size(), 3);

        assertTrue(occurrences.stream().anyMatch(to -> "apple".equals(to.getTerm()) && to.getOccurrenceCount() == 2));
        assertTrue(occurrences.stream().anyMatch(to -> "banana".equals(to.getTerm()) && to.getOccurrenceCount() == 1));
        assertTrue(occurrences.stream().anyMatch(to -> "orange".equals(to.getTerm()) && to.getOccurrenceCount() == 1));
    }

    @Test
    public void testGetDistinctNoteIdCount() {
        // Insert test data
        Long noteId = 1L;
        List<NoteTermFrequency> noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 1));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        noteId = 2L;
        noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "banana", 1));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        noteId = 2L;
        noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "orange", 1));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        assertEquals(2, noteTermFrequencyDao.getDistinctNoteIdCount());
    }

    @Test
    public void testDeleteTermFrequencies() {
        // Insert test data
        Long noteId = 1L;
        List<NoteTermFrequency> noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 2));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "banana", 1));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        // Verify data was inserted
        assertEquals(2, noteTermFrequencyDao.readTermFrequenciesOfNote(noteId).size());

        // Delete data
        noteTermFrequencyDao.deleteTermFrequencies(noteId);

        // Verify data was deleted
        assertTrue(noteTermFrequencyDao.readTermFrequenciesOfNote(noteId).isEmpty());
    }

    @Test
    public void testGetNoteIdsForTerms() {
        // Insert test data
        Long noteId = 1L;
        List<NoteTermFrequency> noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 2));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "banana", 1));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        noteId = 2L;
        noteTermFrequencyList = new ArrayList<>();
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "apple", 1));
        noteTermFrequencyList.add(new NoteTermFrequency(noteId, "orange", 3));
        noteTermFrequencyDao.insertTermFrequenciesToDb(noteTermFrequencyList);

        // Test getNoteIdsForTerms
        Set<String> termsToQuery = new HashSet<>();
        termsToQuery.add("apple");
        termsToQuery.add("banana");

        List<NoteTermFrequency> result = noteTermFrequencyDao.getNoteIdsForTerms(termsToQuery);

        assertTrue(result.stream().anyMatch(to -> "apple".equals(to.getTerm()) && to.getNoteId() == 1));
        assertTrue(result.stream().anyMatch(to -> "apple".equals(to.getTerm()) && to.getNoteId() == 2));
        assertTrue(result.stream().anyMatch(to -> "banana".equals(to.getTerm()) && to.getNoteId() == 1));
    }
}

File: app/src/test/java/com/originb/inkwisenote2/io/sql/NoteTermFrequencyDaoTest.kt
package com.originb.inkwisenote2.io.sql

import androidx.room.Room.inMemoryDatabaseBuilder
import com.originb.inkwisenote2.common.NotesDatabase
import com.originb.inkwisenote2.modules.ocr.data.NoteTermFrequency
import com.originb.inkwisenote2.modules.ocr.data.NoteTermFrequencyDao
import com.originb.inkwisenote2.modules.ocr.data.TermOccurrence
import junit.framework.TestCase.assertEquals
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment

@RunWith(RobolectricTestRunner::class)
class NoteTermFrequencyDaoTest {
    private var noteTermFrequencyDao: NoteTermFrequencyDao? = null
    private var db: NotesDatabase? = null

    @Before
    fun setUp() {
        db = inMemoryDatabaseBuilder<NotesDatabase>(
            RuntimeEnvironment.getApplication(),
            NotesDatabase::class.java
        )
            .allowMainThreadQueries() // Use cautiously, only for tests
            .build()
        noteTermFrequencyDao = db!!.noteTermFrequencyDao()
    }

    @After
    fun tearDown() {
        db!!.close()
    }

    @Test
    fun testInsertAndReadTermFrequencies() {
        // Prepare test data
        val noteId = 1L
        val noteTermFrequencyList: MutableList<NoteTermFrequency> = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 2))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "banana", 1))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "orange", 3))

        // Insert data
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        val termFrequencies: MutableList<NoteTermFrequency> = noteTermFrequencyDao!!.allTermFrequencies
        Assert.assertEquals(noteTermFrequencyList.size.toLong(), termFrequencies.size.toLong())

        for (expected in noteTermFrequencyList) {
            val areTheyEqual = termFrequencies.stream().anyMatch { actual: NoteTermFrequency? ->
                actual!!.noteId == expected.noteId && actual.term == expected.term
                        && actual.termFrequency == expected.termFrequency
            }
            Assert.assertTrue(areTheyEqual)
        }
    }

    @Test
    fun testGetTermOccurrences() {
        // Insert test data for multiple notes
        var noteId = 1L
        var noteTermFrequencyList: MutableList<NoteTermFrequency> = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 2))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "banana", 1))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        noteId = 2L
        noteTermFrequencyList = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 1))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "orange", 3))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        // Test getTermOccurrences
        val termsToQuery: MutableSet<String> = HashSet<String>()
        termsToQuery.add("apple")
        termsToQuery.add("banana")
        termsToQuery.add("orange")

        val occurrences: MutableList<TermOccurrence> = noteTermFrequencyDao!!.getTermOccurrences(termsToQuery)

        Assert.assertEquals(occurrences.size.toLong(), 3)

        Assert.assertTrue(
            occurrences.stream().anyMatch { to: TermOccurrence? -> "apple" == to!!.term && to.occurrenceCount == 2 })
        Assert.assertTrue(
            occurrences.stream().anyMatch { to: TermOccurrence? -> "banana" == to!!.term && to.occurrenceCount == 1 })
        Assert.assertTrue(
            occurrences.stream().anyMatch { to: TermOccurrence? -> "orange" == to!!.term && to.occurrenceCount == 1 })
    }

    @Test
    fun testGetDistinctNoteIdCount() {
        // Insert test data
        var noteId = 1L
        var noteTermFrequencyList: MutableList<NoteTermFrequency> = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 1))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        noteId = 2L
        noteTermFrequencyList = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "banana", 1))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        noteId = 2L
        noteTermFrequencyList = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "orange", 1))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        assertEquals(2, noteTermFrequencyDao!!.distinctNoteIdCount)
    }

    @Test
    fun testDeleteTermFrequencies() {
        // Insert test data
        val noteId = 1L
        val noteTermFrequencyList: MutableList<NoteTermFrequency> = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 2))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "banana", 1))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        // Verify data was inserted
        Assert.assertEquals(2, noteTermFrequencyDao!!.readTermFrequenciesOfNote(noteId).size.toLong())

        // Delete data
        noteTermFrequencyDao!!.deleteTermFrequencies(noteId)

        // Verify data was deleted
        Assert.assertTrue(noteTermFrequencyDao!!.readTermFrequenciesOfNote(noteId).isEmpty())
    }

    @Test
    fun testGetNoteIdsForTerms() {
        // Insert test data
        var noteId = 1L
        var noteTermFrequencyList: MutableList<NoteTermFrequency> = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 2))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "banana", 1))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        noteId = 2L
        noteTermFrequencyList = ArrayList<NoteTermFrequency>()
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "apple", 1))
        noteTermFrequencyList.add(NoteTermFrequency(noteId, "orange", 3))
        noteTermFrequencyDao!!.insertTermFrequenciesToDb(noteTermFrequencyList)

        // Test getNoteIdsForTerms
        val termsToQuery: MutableSet<String> = HashSet<String>()
        termsToQuery.add("apple")
        termsToQuery.add("banana")

        val result: MutableList<NoteTermFrequency> = noteTermFrequencyDao!!.getNoteIdsForTerms(termsToQuery)

        Assert.assertTrue(
            result.stream().anyMatch { to: NoteTermFrequency? -> "apple" == to!!.term && to.noteId == 1L })
        Assert.assertTrue(
            result.stream().anyMatch { to: NoteTermFrequency? -> "apple" == to!!.term && to.noteId == 2L })
        Assert.assertTrue(
            result.stream().anyMatch { to: NoteTermFrequency? -> "banana" == to!!.term && to.noteId == 1L })
    }
}